docs/tutorial.md

```diff
@@ -0,0 +1,829 @@
+
+```ruby
+class BooksController < ApplicationController
+
+  # ...
+
+  # POST /books
+  def create
+    book_action = BA::Book::Create.as(current_user).new(book_params)
+      if book_action.perform
+        redirect_to book_action.subject, notice: 'Book was successfully created.'
+      else
+        @book = book_action.subject
+        render :new
+      end
+    end
+  end
+
+  # ...
+end
+```
+
+## Book::Rent
+
+The Rental system description says:
+
+- All available books can be **rented**
+- Logged users can rent a book
+- A book is not **available** when it's rented to someone
+- A book is **available** after it's delivered back
+
+So, what we're going to do is:
+
+1. Generate migration to create the rental table referencing the book and the user
+2. Add an `available` boolean column in the books table
+3. Create a business action `Book::Rent` and test the conditions above
+
+Let's create `Rent` model first:
+
+```bash
+rails g model rent book:references user:references delivered_back_at:timestamp
+```
+
+and add an `available` column in the books table:
+
+```bash
+rails g migration add_availability_to_books available:boolean
+```
+
+Now it's time to generate the next granite action:
+
+```bash
+rails g granite book/rent
+```
+
+## Preconditions
+
+Let's write specs for the preconditions first:
+
+```ruby
+RSpec.describe BA::Book::Rent do
+  subject(:action) { described_class.as(performer).new(book) }
+
+  let(:performer) { User.new }
+
+  let(:book) { Book.new(title: 'First book', available: available) }
+
+  describe 'preconditions' do
+    context 'with an available book' do
+      let(:available) { true }
+      it { is_expected.to be_satisfy_preconditions }
+    end
+
+    context 'with an unavailable book' do
+      let(:available) { false }
+      it { is_expected.to be_invalid }
+      it { is_expected.not_to satisfy_preconditions }
+    end
+  end
+end
+```
+
+[Preconditions](/granite/#preconditions) are related to the book in the context.
+And the action will decline the context not to be executed if it does not satisfy the preconditions.
+
+Let's implement the `precondition` and `perform` code:
+
+```ruby
+class BA::Book::Rent < BA::Book::BusinessAction
+  precondition { book.available? }
+
+  private
+
+  def execute_perform!(*)
+    Rental.create!(book: subject, user: performer)
+    subject.available = false
+    subject.save!
+  end
+end
+```
+
+Now, let's cover the perform with another spec:
+
+```ruby
+RSpec.describe BA::Book::Rent do
+  subject(:action) { described_class.as(performer).new(book) }
+
+  let(:performer) { User.new }
+
+  let(:book) { Book.new(title: 'First book', available: available) }
+
+  # describe 'preconditions' ...
+
+  describe '#perform!' do
+    specify do
+      expect { action.perform! }
+        .to change(book, :available).from(true).to(false)
+        .and change(Rent, :count).by(1)
+    end
+  end
+end
+```
+
+## Book::DeliverBack
```

Comment:
Maybe `Book::Return` is a better action name

Reply:
Thanks for the suggestion. I'll be renaming it soon.